File: hustack/Ex12_Gold_mining/codes/dp(ver2).py
from collections import deque

if __name__ == '__main__':
    n, L1, L2 = list(map(int, input().split()))
    
    a = list(map(int, input().split()))
    
    dp = [0 for _ in range(n)]
    ans = 0
    dq = deque()
    
    for i in range(n):
        while dq and dq[0] < i - L2:
            dq.popleft()
        
        if i - L1 >= 0:
            j = i - L1
            while dq and dp[dq[-1]] <= dp[j]:
                dq.pop()
            dq.append(j)
        
        best_prev = dp[dq[0]] if dq else 0
        dp[i] = a[i] + best_prev
        ans = max(ans, dp[i])
    
    print(ans)

File: hustack/Ex12_Gold_mining/codes/dp(ver2).cpp
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>
using namespace std;

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, L1, L2;
    cin >> n >> L1 >> L2;
    
    vector<int> a(n);
    for (int i = 0; i < n; i++){
        cin >> a[i];
    }
    
    vector<int> dp(n, 0);
    int ans = 0;
    deque<int> dq;
    
    for (int i = 0; i < n; i++){
        while (!dq.empty() && dq.front() < i - L2)
            dq.pop_front();
        
        if (i - L1 >= 0) {
            int j = i - L1;
            while (!dq.empty() && dp[dq.back()] <= dp[j])
                dq.pop_back();
            dq.push_back(j);
        }
        
        int best_prev = (!dq.empty()) ? dp[dq.front()] : 0;
        dp[i] = a[i] + best_prev;
        ans = max(ans, dp[i]);
    }
    
    cout << ans;
    return 0;
}
